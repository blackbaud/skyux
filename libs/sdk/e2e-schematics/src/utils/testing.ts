import {
  E2eTestRunner,
  UnitTestRunner,
  applicationGenerator,
  libraryGenerator,
} from '@nx/angular/generators';
import { Tree } from '@nx/devkit';
import { Linter } from '@nx/eslint';

export async function createTestApplication(
  tree: Tree,
  options: { name: string; e2eTestRunner?: boolean; unitTestRunner?: boolean },
): Promise<void> {
  await applicationGenerator(tree, {
    name: options.name,
    skipPackageJson: true,
    directory: `apps/${options.name}`,
    e2eTestRunner: options.e2eTestRunner
      ? E2eTestRunner.Cypress
      : E2eTestRunner.None,
    linter: Linter.None,
    unitTestRunner: options.unitTestRunner
      ? UnitTestRunner.Jest
      : UnitTestRunner.None,
  });
}

export async function createTestLibrary(
  tree: Tree,
  options: { name: string },
): Promise<void> {
  await libraryGenerator(tree, {
    name: options.name,
    skipPackageJson: true,
    directory: `libs/${options.name}`,
    linter: Linter.None,
    unitTestRunner: UnitTestRunner.None,
  });
}
