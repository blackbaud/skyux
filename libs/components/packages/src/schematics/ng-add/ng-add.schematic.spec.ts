import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';

import path from 'path';

import { createTestLibrary } from '../testing/scaffold';

const COLLECTION_PATH = path.resolve(__dirname, '../../../collection.json');

describe('ng-add.schematic', () => {
  const runner = new SchematicTestRunner('schematics', COLLECTION_PATH);
  const defaultProjectName = 'my-lib';

  let tree: UnitTestTree;

  beforeEach(async () => {
    tree = await createTestLibrary(runner, {
      name: defaultProjectName,
    });
  });

  function runSchematic(
    options: { project?: string } = {}
  ): Promise<UnitTestTree> {
    return runner.runSchematicAsync('ng-add', options, tree).toPromise();
  }

  it('should apply a fix for crossvent "global is not defined" error', async () => {
    const updatedTree = await runSchematic();

    expect(
      updatedTree
        .readContent('projects/my-lib/src/test.ts')
        .match(/\(window as any\)\.global = window/)?.length
    ).toEqual(1);
  });

  it('should install @angular/cdk', async () => {
    const updatedTree = await runSchematic();

    const packageJson = JSON.parse(updatedTree.readContent('package.json'));

    expect(packageJson.dependencies['@angular/cdk']).toBeDefined();
  });

  it('should install essential SKY UX packages', async () => {
    const updatedTree = await runSchematic();

    const packageJson = JSON.parse(updatedTree.readContent('package.json'));

    const packageNames = [
      '@skyux/assets',
      '@skyux/core',
      '@skyux/i18n',
      '@skyux/theme',
    ];

    for (const packageName of packageNames) {
      expect(packageJson.dependencies[packageName]).toEqual(
        '^0.0.0-PACKAGES_PLACEHOLDER'
      );
    }
  });

  it('should add SKY UX theme stylesheets', async () => {
    const updatedTree = await runSchematic();

    const angularJson = JSON.parse(updatedTree.readContent('angular.json'));

    expect(
      angularJson.projects['my-lib-showcase'].architect.build.options.styles
    ).toEqual([
      'node_modules/@skyux/theme/css/sky.css',
      'node_modules/@skyux/theme/css/themes/modern/styles.css',
      'projects/my-lib-showcase/src/styles.css',
    ]);
  });
});
