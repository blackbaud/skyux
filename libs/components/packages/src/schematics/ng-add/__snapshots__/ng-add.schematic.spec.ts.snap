// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ng-add.schematic should install dependencies 1`] = `
"{
  "dependencies": {
    "@angular/cdk": "^20.0.0",
    "@skyux/assets": "^0.0.0-PLACEHOLDER",
    "@skyux/core": "^0.0.0-PLACEHOLDER",
    "@skyux/i18n": "^0.0.0-PLACEHOLDER",
    "@skyux/theme": "^0.0.0-PLACEHOLDER"
  }
}"
`;

exports[`ng-add.schematic should not update workspace config if a workspace has multiple projects 1`] = `
"{
  "$schema": "./node_modules/@angular/cli/lib/config/schema.json",
  "version": 1,
  "newProjectRoot": "projects",
  "projects": {
    "my-project": {
      "projectType": "library",
      "root": "projects/my-project",
      "sourceRoot": "projects/my-project/src",
      "prefix": "lib",
      "architect": {
        "build": {
          "builder": "@angular/build:ng-packagr",
          "configurations": {
            "production": {
              "tsConfig": "projects/my-project/tsconfig.lib.prod.json"
            },
            "development": {
              "tsConfig": "projects/my-project/tsconfig.lib.json"
            }
          },
          "defaultConfiguration": "production"
        },
        "test": {
          "builder": "@angular/build:karma",
          "options": {
            "tsConfig": "projects/my-project/tsconfig.spec.json"
          }
        }
      }
    },
    "my-project-showcase": {
      "projectType": "application",
      "schematics": {},
      "root": "projects/my-project-showcase",
      "sourceRoot": "projects/my-project-showcase/src",
      "prefix": "app",
      "architect": {
        "build": {
          "builder": "@angular/build:application",
          "options": {
            "browser": "projects/my-project-showcase/src/main.ts",
            "polyfills": [
              "zone.js"
            ],
            "tsConfig": "projects/my-project-showcase/tsconfig.app.json",
            "assets": [
              {
                "glob": "**/*",
                "input": "projects/my-project-showcase/public"
              }
            ],
            "styles": [
              "projects/my-project-showcase/src/styles.css"
            ]
          },
          "configurations": {
            "production": {
              "budgets": [
                {
                  "type": "initial",
                  "maximumWarning": "500kB",
                  "maximumError": "1MB"
                },
                {
                  "type": "anyComponentStyle",
                  "maximumWarning": "4kB",
                  "maximumError": "8kB"
                }
              ],
              "outputHashing": "all"
            },
            "development": {
              "optimization": false,
              "extractLicenses": false,
              "sourceMap": true
            }
          },
          "defaultConfiguration": "production"
        },
        "serve": {
          "builder": "@angular/build:dev-server",
          "configurations": {
            "production": {
              "buildTarget": "my-project-showcase:build:production"
            },
            "development": {
              "buildTarget": "my-project-showcase:build:development"
            }
          },
          "defaultConfiguration": "development"
        },
        "extract-i18n": {
          "builder": "@angular/build:extract-i18n"
        },
        "test": {
          "builder": "@angular/build:karma",
          "options": {
            "polyfills": [
              "zone.js",
              "zone.js/testing"
            ],
            "tsConfig": "projects/my-project-showcase/tsconfig.spec.json",
            "assets": [
              {
                "glob": "**/*",
                "input": "projects/my-project-showcase/public"
              }
            ],
            "styles": [
              "projects/my-project-showcase/src/styles.css"
            ]
          }
        }
      }
    }
  }
}
"
`;

exports[`ng-add.schematic should update workspace config if a workspace only has one project 1`] = `
"{
  "$schema": "./node_modules/@angular/cli/lib/config/schema.json",
  "version": 1,
  "newProjectRoot": "projects",
  "projects": {
    "my-project": {
      "projectType": "application",
      "schematics": {
        "@schematics/angular:component": {
          "style": "scss"
        }
      },
      "root": "",
      "sourceRoot": "src",
      "prefix": "app",
      "architect": {
        "build": {
          "builder": "@angular/build:application",
          "options": {
            "browser": "src/main.ts",
            "polyfills": [
              "zone.js",
              "@skyux/packages/polyfills"
            ],
            "tsConfig": "tsconfig.app.json",
            "inlineStyleLanguage": "scss",
            "assets": [
              {
                "glob": "**/*",
                "input": "public"
              }
            ],
            "styles": [
              "src/styles.scss",
              "@skyux/theme/css/sky.css",
              "@skyux/theme/css/themes/modern/styles.css"
            ],
            "allowedCommonJsDependencies": [
              "@skyux/icons",
              "autonumeric",
              "dom-autoscroller",
              "dragula",
              "fontfaceobserver",
              "intl-tel-input",
              "moment"
            ]
          },
          "configurations": {
            "production": {
              "budgets": [
                {
                  "type": "initial",
                  "maximumWarning": "500kB",
                  "maximumError": "1MB"
                },
                {
                  "type": "anyComponentStyle",
                  "maximumWarning": "4kB",
                  "maximumError": "8kB"
                }
              ],
              "outputHashing": "all"
            },
            "development": {
              "optimization": false,
              "extractLicenses": false,
              "sourceMap": true
            }
          },
          "defaultConfiguration": "production"
        },
        "serve": {
          "builder": "@angular/build:dev-server",
          "configurations": {
            "production": {
              "buildTarget": "my-project:build:production"
            },
            "development": {
              "buildTarget": "my-project:build:development"
            }
          },
          "defaultConfiguration": "development"
        },
        "extract-i18n": {
          "builder": "@angular/build:extract-i18n"
        },
        "test": {
          "builder": "@angular/build:karma",
          "options": {
            "polyfills": [
              "zone.js",
              "zone.js/testing",
              "@skyux/packages/polyfills"
            ],
            "tsConfig": "tsconfig.spec.json",
            "inlineStyleLanguage": "scss",
            "assets": [
              {
                "glob": "**/*",
                "input": "public"
              }
            ],
            "styles": [
              "src/styles.scss",
              "@skyux/theme/css/sky.css",
              "@skyux/theme/css/themes/modern/styles.css"
            ]
          }
        }
      }
    }
  }
}
"
`;
